### DISCUSSION TITLE
Displaying custom errors in the modal window when using flows

### DISCUSSION THREAD

**Discussion by gloribus** [directus/directus#18634] [Feature Requests]
### Summary

Now when using "Run script" in threads, if you want to inform users about an error and terminate the process (filter) throw new Error() is used

But error prints only for administrator. 

```js
            if (req.accountability?.admin === true) {
            payload = {
                errors: [
                    {
                        message: err.message,
                        { extensions: {
                            code: 'INTERNAL_SERVER_ERROR',
                            { err.extensions,
                        },
                    },
                ],
            };
            }
```

I propose to add an error return method from the code, so you can set your own header, text, and error code. Right now an unexpected error is displayed.

You can also add a basic "inform user" operation

### Motivation

Often there are additional checks in the threads and there is no way to report that his request has ended for such and such reason. 

### Detailed Design

There is already a mechanism to display a modal window on error, you need to give the possibility to display it from the flows

### Must Have:
- Method for displaying a modal window with an error from "Run Script" 
- Basic output operation with an error

### Should Have:
- Ability to display not in the modal window, but in the place of validation error messages

### Won't Have:
- A simple message "An unexpected error occurred"

### Drawbacks

I don't think so, because the function is really useful for custom solutions

### Alternatives

Manually change node_module every update or hope that users learn psychic skills 
  

### Adoption Strategy

I think that developers who actively use streams will appreciate this update. The standard error output and its masking for non-administrators will remain, so it doesn't break anything

**Comment by d3v1an7** [comment]
I can't get a modal to appear, even for admins. I really want to throw "something" at the user to let them know the flow failed.
@gloribus does `throw new Error()` still work for you? If so, can you share what I might be missing?

**Reply by gloribus** [reply]
`throw new Error('test message');`
v10.5.3
for admin still work

**Reply by shipsides-sparknel** [reply]
you can take it a step further and use the built in error format, that way instead of displaying the pop up it displays the error in the app page as normal. however, like the normal throw this still only works for admins.

it would be cool if this feature could work in conjunction with this default extension to allow customizable errors that fit the method of others.

``````
if (currentQualDocs.length === 0) { 
                throw {
                    message: 'Qualifications must be uploaded',   
                    extensions: {
                        code: 'FAILED_VALIDATION',
                        field: 'Qualifications_Upload',
                        type: 'nempty'
                    }
                }
} 
``````

**Reply by d3v1an7** [reply]
Unfortunately, neither of these examples above work for me on Directus 10.5.3, even with an admin user.
I only get the temp pop up in the bottom right saying `Flow "Test" ran successfully`, regardless of success or failure.

But I'm using the cloud hosted version -- am wondering if that might be why? I thought the codebase was the same...

![Screenshot 2023-08-29 at 3 40 35 pm](https://github.com/directus/directus/assets/4087799/00b9bce1-0dfe-44f3-bbf0-f2d128669dd4)

![Screenshot 2023-08-29 at 3 40 46 pm](https://github.com/directus/directus/assets/4087799/97e45d47-70a1-45d2-a495-cca110173d6b)

![Screenshot 2023-08-29 at 3 41 03 pm](https://github.com/directus/directus/assets/4087799/188e2f85-8dac-4b89-812c-fcc570f92138)

**Reply by shipsides-sparknel** [reply]
the manual trigger is not a good one to use for this example, it doesn't return async results in the app. you can only see the error results in app in a filter flow.

edit: didn't fully read before posting lol

**Comment by jeffsmonteiro** [comment]
First, I need to say that this feature is super important for better control of flows, and mainly to improve the user experience in the app.

Second, performing tests (as ADMIN), I was able to print the modal by removing the extension parameters from throw (example below). But unfortunately, I can't edit sensitive modal issues like Title, Message Color and Button Text.

```
module.exports = async function(data) {
    throw {
        message: 'Your message from flow here using JS Block',
        extensions: {
            code: 'MYCODE',
        }
    }
    return {};
}
```

Unfortunately too, the Directus documentation also doesn't talk about using throw and its parameters. It would be a great start to improve the documentation.

![image](https://github.com/directus/directus/assets/595398/3bdf17e1-03d6-4246-9bb4-03e55d30ae73)

It's here the CODES list from source:

```
export enum ErrorCode {
	ContainsNullValues = 'CONTAINS_NULL_VALUES',
	ContentTooLarge = 'CONTENT_TOO_LARGE',
	Forbidden = 'FORBIDDEN',
	IllegalAssetTransformation = 'ILLEGAL_ASSET_TRANSFORMATION',
	InvalidCredentials = 'INVALID_CREDENTIALS',
	InvalidForeignKey = 'INVALID_FOREIGN_KEY',
	InvalidIp = 'INVALID_IP',
	InvalidOtp = 'INVALID_OTP',
	InvalidPayload = 'INVALID_PAYLOAD',
	InvalidProvider = 'INVALID_PROVIDER',
	InvalidProviderConfig = 'INVALID_PROVIDER_CONFIG',
	InvalidQuery = 'INVALID_QUERY',
	InvalidToken = 'INVALID_TOKEN',
	MethodNotAllowed = 'METHOD_NOT_ALLOWED',
	NotNullViolation = 'NOT_NULL_VIOLATION',
	RangeNotSatisfiable = 'RANGE_NOT_SATISFIABLE',
	RecordNotUnique = 'RECORD_NOT_UNIQUE',
	RequestsExceeded = 'REQUESTS_EXCEEDED',
	RouteNotFound = 'ROUTE_NOT_FOUND',
	ServiceUnavailable = 'SERVICE_UNAVAILABLE',
	TokenExpired = 'TOKEN_EXPIRED',
	UnexpectedResponse = 'UNEXPECTED_RESPONSE',
	UnprocessableContent = 'UNPROCESSABLE_CONTENT',
	UnsupportedMediaType = 'UNSUPPORTED_MEDIA_TYPE',
	UserSuspended = 'USER_SUSPENDED',
	ValueOutOfRange = 'VALUE_OUT_OF_RANGE',
	ValueTooLong = 'VALUE_TOO_LONG',
}
```

Looking at the source code, I believe it would not be difficult to implement a palliative solution by creating a generic model to address dialog desires from the flow.

**Reply by ched-dev** [reply]
You can create custom errors with [`@directus/errors`](https://docs.directus.io/packages/@directus/errors/) as of Directus 10.4

Is that what you are mentioning?

**Reply by shipsides-sparknel** [reply]
this isn't an option for directus cloud (if i tried to do it right)

**Reply by Malu44** [reply]
Can you use @directus/errors in a flow?

**Reply by DanielBiegler** [reply]
> Can you use @directus/errors in a flow?

As of v10.6 the implementation of the "Run Script"-Operation changed and doesnt allow you to import any other modules.

In a custom operation you can do anything, so for those, yes.

Apart from that the `createError` function doesnt do much, you can just create those yourself, take a look: https://github.com/directus/directus/blob/main/packages/errors/src/create-error.ts

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by jeffsmonteiro** [comment]
Cool!

Given the relevance of the user experience and the complexity that flows can adopt, it would be of great value to offer not only the possibility of displaying the modal, but also offering customized actions for it, which does not seem complex to me given the architecture of the flow itself. Because the modal could be a block and its response (initially with two buttons) could route to the next block.

This can be very important because some actions in the flow can be destructive or impact other data, such as statuses or values in other collections.

In this case, it should also be possible to edit the button names. Their display could continue as the default, a secondary (cancellation/refusal) and a primary (success/confirmation).


![image](https://github.com/directus/directus/assets/595398/28c26586-3230-4fd0-91a6-45fce97c6ce7)

**Comment by rijkvanzanten** [comment]
Related feature request that touches on the implementation for the underlying error message: https://github.com/directus/directus/discussions/19178

**Comment by hauzdigital** [comment]
I'm really looking forward to this feature! As we are currently unable to provide feedback to App users (not admins), and this has generated some frustration.

As already said above, providing continuous feedback to the user is a UX practice.

**Comment by floux-digital** [comment]
OMG! Any news about this? We really need this! In addition, I would love to see what @jeffsmonteiro  commented on. Adding these modals to flows would be amazing and very practical!

**Comment by MarkBurvs** [comment]
Also would love an update on this. It would be amazing to be able to trigger a modal from a flow.

**Comment by robinsonjohn** [comment]
Another vote for the ability to display a modal/custom error message from a flow!

Edit: Modals should be able to appear for any reason (i.e. a success message or purely informational), not only for errors.

**Comment by thomasmaisel** [comment]
Can someone confirm that currently it is not possible to display any kind of error message to a non-admin user from a Webhook/RunScript-Operator except "[INTERNAL_SERVER_ERROR] {}" using only the professional edition.

So I need to buy the enterprise version (minimum: 1000$ a month) and use custom extensions to do that simple task?
I must be missing something, right?

**Reply by shipsides-sparknel** [reply]
seems like a great idea for Directus Labs to pick up as a marketplace extension 😅

**Reply by iksent** [reply]
I tried several options like `throw new Error` and `createError`, but I couldn't get the error to show, only an alert about the successful execution of the script

**Reply by thomasmaisel** [reply]
> I tried several options like `throw new Error` and `createError`, but I couldn't get the error to show, only an alert about the successful execution of the script

Yes, that's the problem.

**Comment by mehunk** [comment]
Any update about this? It is already 2025...

**Reply by gloribus** [reply]
> Any update about this? It is already 2025...

Forget about such functions. It's sad but they are moving in the wrong direction

**Comment by d3v1an7** [comment]
FYI for anyone following along, this is a thing as of [a few weeks back](https://github.com/directus/directus/pull/24965) and looks like it shipped with [v11.7.0](https://github.com/directus/directus/milestone/218). There is a new option 'Error on Reject' when setting up the flow, which is off by default.
![Screenshot 2025-05-01 at 11 24 01 pm](https://github.com/user-attachments/assets/7b806f54-e84e-4cef-b757-97b428090059)

After enabling this, any throw message in the flow will end up in an error modal.

My context/usage for this so far is with Manually triggered flows, where a user presses a button and the flow runs.

From testing, all of these work:
```
throw 'Error message';
throw { message: 'Error message' };
throw new Error('Error message');
```

The user sees a modal like this:
![Screenshot 2025-05-01 at 11 30 16 pm](https://github.com/user-attachments/assets/2c83b07b-7fb4-43fc-8093-b8fb5f5560a1)

I couldn't find a way to change/update the modal title (Unexpected Error) or the error code (INTERNAL_SERVER_ERROR), but I didn't dig too far into the source. Regardless, perfect is the enemy of good :)

**Comment by Abdallah-Awwad** [comment]
With version 11.10.1 a new [operation released](https://github.com/directus/directus/pull/25558) for throwing errors, although in the run script it still doesn't work, but now at least there is a dedicated operation for it.

**Comment by hermanndettmann** [comment]
I have kind of the same issue but from another angle as I have some filter event hooks running where I throw some errors with `throw new Error()` and realised they are only visible to admin users. It would be very helpful to have a way to throw errors in this context that are visible to all users instead of just showing them a generic message.

**Reply by MarkBurvs** [reply]
You should be able to use the new Throw Error operation instead (added in 11.10.1) and that will show okay to non-admin users.

**Reply by hermanndettmann** [reply]
Thanks @MarkBurvs for your reply! But how should I use an operation in a [filter event hook extension](https://directus.io/docs/guides/extensions/api-extensions/hooks)?

**Reply by MarkBurvs** [reply]
Oh I see sorry I didn't realise you're talking about extensions rather than Flows. I don't know then sorry!

**Reply by hermanndettmann** [reply]
Alright. Thanks anyway! Maybe my comment wasn't precise enough in this context.
